### PR TITLE
WIP: Competence based training

### PR DESCRIPTION
**Pull request by jsenellart**
This PR is introducing "Competence Based Curriculum Learning" (https://arxiv.org/pdf/1903.09848.pdf).
It seems to be functional, and I implemented both "linear" and "sqroot" competence simulation. Currently in testing on large training and needs documentation - @guillaumekln : could you please look at the code and let me know your thoughts regarding the implementation choice? Thanks.

### CODE REVIEW

**Comment by guillaumekln** [comment]
The overall integration looks good to me. However, I noted that it would be preferable for `training_pipeline` to remain generic, for example:

* zip the dataset before calling the function
* add a `filter_fn` argument that is called after shuffling
* rely on `process_fn` to remove the extra value

This would also fix the case `num_shards > 1` which does not work with the current implementation. 

Let me know how the experiment is going. I will have more detailed comments if improvements are measurable and we want to get this in.

File: opennmt/data/competence.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import tensorflow as tf
```
Comment (guillaumekln) [inline opennmt/data/competence.py:1]: Looks like this file can be removed.

Comment (jsenellart) [reply opennmt/data/competence.py]: yes - done

**Comment by jsenellart** [comment]
Thanks @guillaumekln : regarding your suggestion for moving the code away from `training_pipeline`, do I correctly understand that you suggest putting the logic inside `make_training_dataset` - by first zipping before the call, +passing a `process_fn` removing the score. Also adding a new `filter_fn`parameter performing the filter logic? For the latter, if we do so, wouldn't it make sense then to use the same code for the current length+ filtering?

**Comment by guillaumekln** [comment]
Yes, that's what I had in mind. `filter_fn` would be a generalization of the current length filtering logic. In this case, callers of `training_pipeline` would be updated to use `filter_fn` instead of the direct maximum length arguments.

**Comment by jsenellart** [comment]
actually, keeping `process_fn` (as a map), and `filter_fn` would not be enough because in CL case, we do need to filter **before** process, and the other filters need to apply **after**. I propose to simply replace `process_fn` by a `process_fns`: an ordered list of process function, that can be either map or filter transformations.

```python
    for process_fn in process_fns or []:
      dataset = dataset.apply(process_fn)
```

**Comment by guillaumekln** [comment]
That makes sense and is very generic.

**Comment by jsenellart** [comment]
what about this? I left the batch filtering functions inside dataset.py since they are more training specific - and moved out the example filtering in the list of process_fns

**Comment by guillaumekln** [comment]
This should work.

**Comment by guillaumekln** [comment]
An analysis of this feature is available on the forum: https://forum.opennmt.net/t/curriculum-learning-in-the-age-of-transformers-parts-i-ii/3648

**Pull request closed without merging**